### DISCUSSION TITLE
How do I schedule a daily partitioned job to run at T instead of T+1?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17696] [Q&A]
hey folks - how do I schedule a daily partitioned job to run at T instead of T+1? e.g., partition 2023-01-01 runs at 2023-01-01 midnight and not 2023-01-02 midnight

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1699043688495349?thread_ts=1699043688.495349&cid=C01U954MEER).*

**Comment by sryza** [comment]
You can accomplish this by putting an `end_offset` on your `DailyPartitionsDefinition`:

```python
from dagster import (
    DailyPartitionsDefinition,
    Definitions,
    asset,
    build_schedule_from_partitioned_job,
    define_asset_job,
)


@asset(partitions_def=DailyPartitionsDefinition(start_date="2023-01-01", end_offset=1))
def asset1():
    ...


scheduled_job = build_schedule_from_partitioned_job(define_asset_job("job1", selection=[asset1]))

defs = Definitions(assets=[asset1], schedules=[scheduled_job])
```

### Explanation of why this works

1. Each partition in a `DailyPartitionsDefinition` corresponds to a time window that starts at the beginning of the date named in the partition key, and ends at the start of the next day.
2. By default, at any time T, the set of partitions that exists in the `DailyPartitionsDefinition` includes all time windows that end before T.
3. The schedule created by `build_schedule_from_partitioned_job` launches a run for the latest partition that exists at the time of the schedule tick.

By setting `end_offset=1` on the `DailyPartitionsDefinition`, you're modifying (2): at any time T, the set of partitions that exists in the `DailyPartitionsDefinition` includes all time windows that end before T, and one more additional daily time window that ends after T.

So when the schedule ticks at time T, it will see that a partition exists whose window ends after T and launch that one.

**Reply by rob-apella** [reply]
@sryza thank you for the explanation. If I set `end_offset` to an existing `DailyPartitionsDefinition`, will Dagster preserve the status history of all partitions that have run thus far? Or will that history get erased?

**Reply by sryza** [reply]
Dagster will preserve the status history

**Reply by rob-apella** [reply]
Great - thank you!!
